### PR TITLE
Introduced javax module for running arbitrary java apps in an Ansible-like way

### PR DESCRIPTION
**Pull request by tima**

### CODE REVIEW

File: commands/javax.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    (rc, out, err) = module.run_command(cmd)
+    if rc:
+        return module.fail_json(msg=err, rc=rc, cmd=cmdstr)
+    else:
+        return module.exit_json(changed=changed, msg=out, rc=rc,
+                                cmd=cmdstr)
+
+# import module snippets
+
+from ansible.module_utils.basic import *
```
Comment (juliedavila) [inline commands/javax.py:167]: This is ideally

``` python
if __name__ == '__main__':
    main()
```

Comment (bcoca) [inline commands/javax.py:167]: not really as modules are not importable due to the 'fake import' from module_utils

File: commands/javax.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  properties:
+    Name: Foo
+    Hello: World
+  classpath:
+    - /path/to/some.jar
+  java_opts:
+    - '-XX:MaxPermSize=512m'
+    - '-XX:+HeapDumpOnOutOfMemoryError'
+'''
+
```
Comment (juliedavila) [inline commands/javax.py:94]: Would be ideal to see a `RETURN` section here. For reference: https://github.com/ansible/ansible-modules-core/blob/devel/packaging/os/apt.py#L141-L161

File: commands/javax.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = \
```
Comment (juliedavila) [inline commands/javax.py:19]: Line continuation isn't really needed here

File: commands/javax.py (hunk 1 above)
Comment (juliedavila) [inline commands/javax.py:167]: This was based on @abadger's commentary on other modules

Comment (abadger) [inline commands/javax.py:167]: Yeah -- at some point in the future we may get unittests hooked up for module code.  If we get that then if **name** == '**main**' is necessary.  (there's already unittesting of a few specific module cases... a general framework would need even more work).

**Comment by abadger** [comment]
bcoca, jimi-c. and I had a lengthy discussion about this yesterday.  We decided that we might not accept this due to duplicating existing ansible functionality.

If you can show us how this is different from using similar data structures along with jinja2 filters and the command module it would let us evaluate whether this is different enough to warrant its own module.  Thanks!

**Comment by tima** [comment]
I'm doing this though I think this misses the point of this module. The purpose of this module is ease of use and readability of playbooks making complex java app calls, not because it can't be done.

That said, this static java execution task:

```
command: java  -Xms1024m -Xmx2048m -classpath "/tmp/javax/test/resources/propsdump.jar" -DHello=World -DName=Foo -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError PropsDump
```

Could be done with some vars and a Jinja templating like this:

```
  vars:
    path: "/tmp/javax/test/resources"
    properties:
      Name: Foo
      Hello: World
    classpath:
      - "{{ path }}/propsdump.jar"
    java_opts:
      - '-XX:MaxPermSize=512m'
      - '-XX:+HeapDumpOnOutOfMemoryError'
  tasks:
    command: java  -Xms1024m -Xmx2048m -classpath "{{ classpath | join(';') }}" {% for k, v in properties.iteritems() %}-D{{k}}={{v}} {% endfor %} {{ java_opts | join(' ') }} PropsDump
```

Using the same vars section above and this module the same call is reproduced with:

```
      javax: javaclass='PropsDump' init_heap_size=1024m max_heap_size=2048m
      args:
        properties: "{{ properties }}"
        classpath: "{{ classpath }}"
        java_opts: "{{ java_opts }}"
```

I could have totally folded the args section in to one big var structure but was being a bit lazy. The thing this examples do not demonstrate is all the conditional logic javax has that the jinja2 templatized module does not.

**Comment by abadger** [comment]
Probably easier to read like this:

```
  vars:
    path: "/tmp/javax/test/resources"
    properties:
      - '-DName=Foo'
      - '-DHello=World'
    classpath:
      - "{{ path }}/propsdump.jar"
    java_opts:
      - '-Xms1024m'
      - '-Xmx2048m'
      - '-XX:MaxPermSize=512m'
      - '-XX:+HeapDumpOnOutOfMemoryError'
    propsdump_java_opts:  "{{ java_opts + properties }}"
  tasks:
    - command: java -classpath "{{ classpath | join(';') }}" {{propsdump_java_opts| join(' ') }} PropsDump
```

and

```
  vars:
    path: "/tmp/javax/test/resources"
    properties:
      Name: Foo
      Hello: World
    classpath:
      - "{{ path }}/propsdump.jar"
    java_opts:
      - '-XX:MaxPermSize=512m'
      - '-XX:+HeapDumpOnOutOfMemoryError'
    tasks:
      - javax:
          javaclass: 'PropsDump'
          init_heap_size: 1024m
          max_heap_size: 2048m
          properties: "{{ properties }}"
          classpath: "{{ classpath }}"
          java_opts: "{{ java_opts }}"

```

Not seeing a whole lot of difference in these two that would justify writing a module here.  What's the conditional logic difference that you talk about?

**Comment by resmo** [comment]
I tend to -1 on this module.

My question is, why should I use this module? If we look other "command" wrapper modules, e.g. the puppet module, there are benefits, e.g. check mode support, better error handling, etc. 

## Pro:
- it makes the use of java execution more ansible like.

## Con:
- Why use it? I see no benefit.
- Can be replaced with existing functionality
- Then we also accept a php exec module? ruby exec module? python exec module?

**Comment by gregdek** [comment]
Adding new process. We will be evaluating all new module PRs according to this process, effective immediately.

# 

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

**Comment by gregdek** [comment]
@tima I interpret @resmo's remarks as "needs_revision".

If you choose to make the revisions, feel free to make them and let us know by putting "ready_for_review" and we will continue with the review process.  Thanks!

**Comment by gregdek** [comment]
@tima A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

**Comment by resmo** [comment]
@tima @gregdek you can ignore my comment since I do not have any "power" to reject this module. Just give my 2 cents.

**Comment by robynbergeron** [comment]
@tima Did you want to keep pursuing this? :) It looks like you were thinking of doing a few of the things @abadger recommended above, but not sure where it stands.

**Comment by robynbergeron** [comment]
@tima Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If we don't hear from you within another 14 days, we will close this pull request.

**Comment by gregdek** [comment]
Closing due to contributor inactivity.

**Pull request closed without merging**